File: Interview Preparation Kit - CPP/07. Search/001. Hash Tables Ice Cream Parlor.cpp
// Problem: https://www.hackerrank.com/challenges/ctci-ice-cream-parlor/problem
// Score: 35


#include <iostream>
#include <map>
using namespace std;


int main(){
    int t = 0;
    cin >> t;

    for (int i = 0; i < t; i++){
        int money, n;
        cin >> money >> n;
        map<int, int> arr;

        for (int j = 0; j < n; j++){
            int c;
            cin >> c;

            if (arr.find(c) == arr.end()){
                arr[money - c] = j + 1;
            }
            else{
                cout << arr[c] << " " << j + 1 << '\n';
            }
        }
    }
    return 0;
}

File: Interview Preparation Kit - Python/07. Search/001. Hash Tables - Ice Cream Parlor.py
# Problem: https://www.hackerrank.com/challenges/ctci-ice-cream-parlor/problem
#  Score: 35


t = int(input())
for i in range(t):
    money = int(input())
    n = int(input())
    arr = list(map(int, input().split()))

    saved_values = {}
    for counter, value in enumerate(arr):
        if money-value in saved_values:
            print(saved_values[money-value] + 1, counter + 1)
        elif value not in saved_values:
            saved_values[value] = counter
